### ISSUE TITLE
Problem private ECR

### ISSUE THREAD

**Issue by sjoerd-dijkstra** [npalm/terraform-aws-gitlab-runner#234]
We're having problems using images from our private ECR. Followed #189 to no avail, the issue seems in instantiating the runners. `amazon-ecr-credential-helper` does not seem to do anything.

In cloud watch we find many log lines:
> gitlab-runner: Failed to process runner  builds error: failed to update executor: no free machines that can process builds

(striped from color codes)

Anybody an idea what the issue could be? We're using docker+machine with spot instances and the module seems to work as expected (registering runners, running jobs, removing spot instances, etc.), except when using private ECR docker images. Gitlab logs keep giving:

> ERROR: Preparation failed: Error response from daemon: Get https://XXXXXX.dkr.ecr.eu-west-1.amazonaws.com/v2/DOCKER-IMAGE-NAME/manifests/latest: no basic auth credentials (docker.go:200:0s)

module config:

```terraform
data "aws_security_group" "default" {
  name   = "default"
  vpc_id = module.vpc.vpc_id
}

module gitlab_runner {
  # https://registry.terraform.io/modules/npalm/gitlab-runner/aws/
  source = "npalm/gitlab-runner/aws"

  aws_region  = local.aws_region
  environment = "gitlab-runners"

  vpc_id                   = module.vpc.vpc_id
  subnet_ids_gitlab_runner = module.vpc.private_subnets
  subnet_id_runners        = element(module.vpc.private_subnets, 0)

  runners_name       = "gitlab-runner"
  runners_gitlab_url = "https://gitlab.com"

  enable_runner_ssm_access = false

  userdata_pre_install     = "sudo yum install amazon-ecr-credential-helper -y"
  userdata_post_install    = "echo '#!/bin/bash\nsudo apt-get update\nsudo apt-get install amazon-ecr-credential-helper -y' > /var/tmp/user_data.sh"
  docker_machine_options   = ["amazonec2-userdata=/var/tmp/user_data.sh"]
  runners_environment_vars = ["DOCKER_AUTH_CONFIG={\"credHelpers\":{\"${local.aws_account_id}.dkr.ecr.${local.aws_region}.amazonaws.com\": \"ecr-login\"}}"]

  ami_filter = {
    name = ["amzn2-ami-hvm-2.0.20200304.0-x86_64-ebs"]
  }

  runner_ami_filter = {
    name = ["ubuntu/images/hvm-ssd/ubuntu-eoan-19.10-amd64-server-20200325"]
  }

  gitlab_runner_security_group_ids = [data.aws_security_group.default.id]

  runners_idle_time             = 600
  runners_idle_count            = 1
  runners_off_peak_timezone     = "Europe/Amsterdam"
  docker_machine_instance_type  = "t3.medium"

  overrides = {
    name_sg=""
    name_runner_agent_instance  = "runner-agent"
    name_docker_machine_runners = "runner"
  }

  gitlab_runner_registration_config = {
    registration_token = local.gitlab_registration_token
    tag_list           = "docker"
    description        = "Spot Runners"
    locked_to_project  = "false"
    run_untagged       = "true"
    maximum_timeout    = "3600"
  }

  tags = local.tags

  runners_image = "docker:19.03"
}

resource aws_iam_role_policy_attachment allow_ecr_full_agent {
  role       = module.gitlab_runner.runner_agent_role_name
  policy_arn = "arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryFullAccess"
}

resource aws_iam_role_policy_attachment allow_ecr_full {
  role       = module.gitlab_runner.runner_role_name
  policy_arn = "arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryFullAccess"
}

```

**Comment by npalm** [comment]
We run the runner for a large set of projects, so we do not grant the runner any extra rights in AWS. Not sure what you exact needs are, but you could create an Policy in AWS that allows access to your private ECR. Next you can attach the policy to the role of the runner `runner_role_name`. You runners should then be able to access ECR. No aws credential help needed as for I know.

**Comment by kayman-mk** [comment]
I have seen the first error message in my logs if the var.aws_zone (default is "a") didn't match the availability_zone of var.subnet_id_runners. See #232 

I guess this is not the problem, but the second error message looks like a missing DOCKER_AUTH variable in the CI system.

**Comment by npalm** [comment]
@sjoerd-dijkstra I hope your issue is resolved, shall we close the issue?

**Comment by bonza** [comment]
Could it be related to [this](https://gitlab.com/gitlab-org/gitlab-runner/-/issues/4426) issue

**Comment by sjoerd-dijkstra** [comment]
Thanks all for the help. Meanwhile we have went with EKS to host our runners. Read permission is only granted from the ECR side, and no extra policies are needed on the EC2 instances. We've configured the actual runners with `amazon-ecr-credential-helper` and the file `~/.docker/config.json` (also described in the issue mentioned by @bonza). No DOCKER_AUTH variable is set, as that part is covered by the file. This configuration should also work in the docker+machine setup, however we were in lack of time to work this out and found another way to host the runners with an existing EKS cluster.

@npalm i will close the issue, thanks again for the help and the project 👍
